broker/fragment/store_azure.go

```diff
@@ -222,47 +237,133 @@ func getAzureStorageCredential(coreCredential azcore.TokenCredential, tenant str
 	return credential, nil
 }
 
-func (a *azureBackend) azureClient(ep *url.URL) (cfg AzureStoreConfig, client pipeline.Pipeline, err error) {
-	if err = parseStoreArgs(ep, &cfg); err != nil {
+func parseAzureEndpoint(endpoint *url.URL) (cfg azureStoreConfig, err error) {
+	if err = parseStoreArgs(endpoint, &cfg); err != nil {
 		return
 	}
+
 	// Omit leading slash from URI. Note that FragmentStore already
 	// enforces that URL Paths end in '/'.
-	var splitPath = strings.Split(ep.Path[1:], "/")
+	var splitPath = strings.Split(endpoint.Path[1:], "/")
 
-	if ep.Scheme == "azure" {
+	if endpoint.Scheme == "azure" {
 		// Since only one non-ad "Shared Key" credential can be injected via
 		// environment variables, we should only keep around one client for
 		// all `azure://` requests.
 		cfg.accountTenantID = "AZURE_SHARED_KEY"
 		cfg.storageAccountName = os.Getenv("AZURE_ACCOUNT_NAME")
-		cfg.containerName, cfg.prefix = ep.Host, ep.Path[1:]
-	} else if ep.Scheme == "azure-ad" {
-		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = ep.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
+		cfg.containerName, cfg.prefix = endpoint.Host, endpoint.Path[1:]
+	} else if endpoint.Scheme == "azure-ad" {
+		cfg.accountTenantID, cfg.storageAccountName, cfg.containerName, cfg.prefix = endpoint.Host, splitPath[0], splitPath[1], strings.Join(splitPath[2:], "/")
 	}
 
-	a.clientMu.Lock()
-	defer a.clientMu.Unlock()
+	return cfg, nil
+}
 
-	if a.clients[cfg.accountTenantID] != nil {
-		client = a.clients[cfg.accountTenantID]
-		return
+func (a *azureBackend) getAzureServiceClient(endpoint *url.URL) (client *service.Client, err error) {
+	var cfg azureStoreConfig
+
+	if cfg, err = parseAzureEndpoint(endpoint); err != nil {
+		return nil, err
 	}
 
-	var credentials azblob.Credential
+	a.mu.Lock()
```

Comment:
nit: can you arrange so that any constructed instance of `azureBackend` already has all maps initialized? Then you don't need to worry about their being nil.

For example:
```
&azureBackend {
   foo: make(map[string]*service.Client),
   bar: make(map[string]string),
}
```

then you can omit nil checks everywhere.

Reply:
Done! It was only constructed in one place, so easy to update.